README.md
<p align="center"><img src="https://cldup.com/Yv3u-lqTHO.png"></p>

![](http://41.media.tumblr.com/6fd06309c8a378665f1d6094925d76b6/tumblr_no8x76bInz1qz6oq7o5_1280.png)
<p align="center"><a href="https://github.com/octalmage/minum/releases/latest"><img width="300" src="https://cldup.com/xVA9LNTaCb.png"></a><a href="http://octalmage.com/tagged/minum"><img width="300" src="https://cldup.com/IkDeam5PrL.png"></a></p>

## Description

minum is a minimalistic web browser that removes all distractions and allows the user to focus on what's important: the website.

I've always loved the way browsers are represented in portfolios. Their job is to showcase the website and stay out of the way. I wanted to recreate that with minum.

## Instructions

Click the website's title to view the address bar, type in a URL or search term, then press enter to go!

![](https://cldup.com/YUGXUr093Q.gif)

## Screenshots

![](http://40.media.tumblr.com/d11eb3efce518e5bc44b936643b9bbc2/tumblr_no8x76bInz1qz6oq7o2_1280.png)

![](http://41.media.tumblr.com/d42c420e0780e78fbf727db839a5d31d/tumblr_no8x76bInz1qz6oq7o3_1280.png)

![](http://41.media.tumblr.com/98fd3920e98d5f5e1c430cef9c8a10c5/tumblr_no8x76bInz1qz6oq7o4_1280.png)

![](http://40.media.tumblr.com/641e343d918ad4551543db4bba450fdd/tumblr_no8x76bInz1qz6oq7o1_1280.png)

## Plans
* Multiple tabs.
* Built in screenshot  tool.
* Delicious bookmark integration.
* Extensions (CSS and JavaScript injection).

## License
GPL v2


CHANGELOG.md
# Changelog

## v0.1.1 (21/06/2018)
- [**bug**] Fixed jumpy scrollbars.  [#9](https://github.com/octalmage/minum/issues/9)
- [**enhancement**] Use Electron.  [#7](https://github.com/octalmage/minum/issues/7)
- [**enhancement**] Use Chrome's webview instead of an iFrame. [#5](https://github.com/octalmage/minum/issues/5)
- [**bug**] Fix copy and paste.  [#11](https://github.com/octalmage/minum/issues/11)


main.js
// eslint-disable-next-line import/no-extraneous-dependencies
const { app, BrowserWindow, Menu } = require('electron');
const { autoUpdater } = require('electron-updater');

autoUpdater.checkForUpdatesAndNotify();

let mainWindow;

function createWindow() {
  mainWindow = new BrowserWindow({
    title: 'minum',
    frame: false,
    width: 1080,
    height: 600,
    show: false,
    backgroundColor: '#ffffff',
  });

  mainWindow.loadFile('index.html');

  mainWindow.on('closed', () => {
    mainWindow = null;
  });

  mainWindow.once('ready-to-show', () => {
    mainWindow.show();
  });

  const template = [{
    label: 'minum',
    submenu: [
      { label: 'About minum', selector: 'orderFrontStandardAboutPanel:' },
      { type: 'separator' },
      { label: 'Quit', accelerator: 'Command+Q', click: () => { app.quit(); } },
    ],
  }, {
    label: 'Edit',
    submenu: [
      { label: 'Undo', accelerator: 'CmdOrCtrl+Z', selector: 'undo:' },
      { label: 'Redo', accelerator: 'Shift+CmdOrCtrl+Z', selector: 'redo:' },
      { type: 'separator' },
      { label: 'Cut', accelerator: 'CmdOrCtrl+X', selector: 'cut:' },
      { label: 'Copy', accelerator: 'CmdOrCtrl+C', selector: 'copy:' },
      { label: 'Paste', accelerator: 'CmdOrCtrl+V', selector: 'paste:' },
      { label: 'Select All', accelerator: 'CmdOrCtrl+A', selector: 'selectAll:' },
    ],
  }];

  Menu.setApplicationMenu(Menu.buildFromTemplate(template));
}

app.on('ready', createWindow);

app.on('window-all-closed', () => {
  app.quit();
});

app.on('activate', () => {
  if (mainWindow === null) {
    createWindow();
  }
});


renderer.js
// eslint-disable-next-line import/no-extraneous-dependencies
const { remote } = require('electron');
const tld = require('tldjs');

const win = remote.getCurrentWindow();
$(() => {
  // Close button.
  $('#closeButton').on('mousedown', () => {
    win.close();
  });

  // Minimize button.
  $('#minButton').on('mousedown', () => {
    win.minimize();
  });

  // Maximize button.
  $('#fullButton').on('mousedown', () => {
    win.setFullScreen(!win.isFullScreen());
  });

  $('#title').on('mousedown', () => {
    $('#info').show(5, () => {
      $('#address').focus();
      $('#address').select();
    });
  });

  // Keydown in address bar.
  $('#address').on('keydown', (e) => {
    // Enter is pressed.
    if (e.which === 13) {
      // Hide the info page.
      $('#info').hide();

      // Get the value of the address bar.
      let search = $('#address').val();

      // If top level domain exists, and there is text on both sides of the dot.
      if (tld.tldExists(search) && search.split('.').length > 1 && search.split('.')[0].length > 0) {
        // Add http to the url if it's missing.
        if (!/^https?/ig.test(search)) {
          search = `http://${search}`;
        }

        $('#title').text(search);
      } else {
        $('#title').text(search);
        search = `https://www.google.com/search?q=${search}`;
      }

      $('#view').attr('src', search);
    }
  });

  // If exit is pressed leave info screen.
  $('#info').on('keydown', (e) => {
    if (e.which === 27) {
      $('#info').hide();
    }
  });

  // Leave info screen if background is clicked.
  $('#info').on('mousedown', (e) => {
    if (e.target.id === 'info') {
      $('#info').hide();
    }
  });
});

function idealTextColor(components) {
  const nThreshold = 105;
  const bgDelta = (components.R * 0.299) + (components.G * 0.587) + (components.B * 0.114);

  return ((255 - bgDelta) < nThreshold) ? '#000000' : '#ffffff';
}

function getBarColor() {
  document.querySelector('webview').capturePage((img) => {
    const canvas = document.getElementById('color');
    const context = canvas.getContext('2d');
    const image = document.getElementById('colorImage');
    image.src = img.toDataURL();

    image.onload = () => {
      // Reset the canvas before drawing our image.
      context.clearRect(0, 0, canvas.width, canvas.height);
      // Draw the image to our invisible canvas.
      context.drawImage(image, 0, 0);

      // Get the pixel color right below the titlebar.
      let pixelData = context.getImageData(0, 0, 1, 1).data;

      // If background is transparent it will appear white so make the titlebar match.
      if (pixelData.every((v, i) => v === [0, 0, 0, 0][i])) {
        pixelData = [255, 255, 255, 255];
      }

      // Change the title bar background color to the color found above.
      document.getElementById('bar').style.backgroundColor = `rgb(${[pixelData[0], pixelData[1], pixelData[2]].join(',')})`;

      // Get the color the title bar text should be based on the title bar background.
      // Currently black or white.
      const inv = idealTextColor({ R: pixelData[0], G: pixelData[1], B: pixelData[2] });

      // Set the title bar text.
      document.getElementById('title').innerHTML = document.querySelector('webview').getTitle();

      // Set the title bar text color.
      document.getElementById('title').style.color = inv;

      // Set the address bar url to the current url.
      $('#address').val(document.querySelector('webview').getURL());
    };
  });
}

document.querySelector('webview').addEventListener('dom-ready', () => {
  getBarColor();
});

document.querySelector('webview').addEventListener('new-window', (event) => {
  $('#view').attr('src', event.url);
});


Gruntfile.js
module.exports = function(grunt)
{

	grunt.initConfig(
	{
		pkg: grunt.file.readJSON("package.json"),
		jshint:
		{
			files: ["main.js", "Gruntfile.js"],
			options:
			{
				globals:
				{
					jQuery: true,
					console: true,
					module: true,
					document: true
				}
			}
		},
		jsbeautifier:
		{
			files: ["main.js", "Gruntfile.js"],
			options:
			{
				js:
				{
					braceStyle: "expand",
					indentWithTabs: true,
					indentSize: 1,
					maxPreserveNewlines: 2
				}
			}
		},
		nodewebkit:
		{
			options:
			{
				platforms: ["osx64"],
				version: "0.12.1",
				buildDir: "./build",
                macIcns: "./logo.icns"
			},
			src: ["index.html", "package.json", "main.js", "node_modules/tldjs/**", "app/**"]
		}
	});

	grunt.loadNpmTasks("grunt-contrib-jshint");
	grunt.loadNpmTasks("grunt-jsbeautifier");
	grunt.loadNpmTasks('grunt-node-webkit-builder');

	grunt.registerTask("test", ["jshint"]);
	grunt.registerTask("clean", ["jsbeautifier"]);
	grunt.registerTask("build", ["nodewebkit"]);
	grunt.registerTask("default", ["jshint"]);

};

mainOld.js
var gui = require("nw.gui");
var tld = require("tldjs");

var win = gui.Window.get();

win.focus();

if (process.platform === "darwin")
{
	var nativeMenuBar = new gui.Menu(
	{
		type: "menubar"
	});
	nativeMenuBar.createMacBuiltin("minum");
	win.menu = nativeMenuBar;
}

//debug
//win.showDevTools();

//App loaded.
$(function()
{
	//Close button.
	$("#closeButton").on("mousedown", function()
	{
		gui.App.quit();
	});	
	
	//Minimize button.
	$("#minButton").on("mousedown", function()
	{
		win.minimize();
	});
	
	//Maximize button.
	$("#fullButton").on("mousedown", function()
	{
		win.maximize();
	});
	
	$("#title").on("mousedown", function()
	{
		$("#info").fadeIn(null, function()
		{
			$("#address").focus();
			$("#address").select();
		});
	});
	
	//Keydown in address bar.
	$("#address").on("keydown", function (e) 
	{
		//Enter is pressed.
  		if (e.which == 13) 
		{
			//Hide the info page. 
    		$("#info").hide();

			//Get the value of the address bar.
			var search = $("#address").val();

			//If top level domain exists, and there is text on both sides of the dot.
			if (tld.tldExists(search) && search.split(".").length > 1 && search.split(".")[0].length > 0)
			{
				//Add http to the url if it's missing.
				if (!/^https?/ig.test(search))
				{
					search = "http://" + search;
				}
				
				$("#title").text(search);	
			}
			else
			{
				$("#title").text(search);
				search = "https://www.google.com/search?q=" + search;
			}
			
			$("#view").attr("src", search);
  		}
	});
	
	//If exit is pressed leave info screen.
	$("#info").on("keydown", function(e)
	{
		if (e.which === 27)
		{
			$("#info").hide();
		}
	});

	//Leave info screen if background is clicked.
	$("#info").on("mousedown", function(e)
	{
		if (e.target.id === "info")
		{
			$("#info").hide();
		}
	});
});

//Force links that target _blank to open in the same frame (until tabs are implemented?).
win.on("new-win-policy", function(frame, url, policy)
{
	policy.forceCurrent();
});

win.on("document-end", function(frame) 
{
    if (frame && frame.hasOwnProperty("id") && frame.id === "view")
    {
        setTimeout(function()
        {
            getBarColor();
        },100);
    }
});

function getBarColor()
{
    win.capturePage(function(img) 
    {   
        var canvas = document.getElementById("color").getContext("2d");
        var image = document.getElementById("colorImage");
        
        image.src = img;
        
        //Draw the image to our invisible canvas. 
        canvas.drawImage(image,0,0);
        
        //Get the pixel color right below the titlebar. 
        var pixelData = canvas.getImageData(5, 35, 1, 1).data;
        
        //Change the title bar background color to the color found above. 
		document.getElementById("bar").style.backgroundColor = 'rgb(' + [pixelData[0],pixelData[1],pixelData[2]].join(',') + ')';

		//Get the color the title bar text should be based on the title bar background. Currently black or white. 
        var inv = idealTextColor({R: pixelData[0], G: pixelData[1], B: pixelData[2]});
        
        //Set the title bar text.
        document.getElementById("title").innerHTML = document.getElementById("view").contentDocument.title;

        //Set the title bar text color. 
        document.getElementById("title").style.color = inv;
		
		//Set the address bar url to the current url.
		$("#address").val($("#view")[0].contentWindow.location.href);
        
    }, "png");
}

function idealTextColor(components) 
{

   var nThreshold = 105;
   var bgDelta = (components.R * 0.299) + (components.G * 0.587) + (components.B * 0.114);

   return ((255 - bgDelta) < nThreshold) ? "#000000" : "#ffffff";   
}